quadratichq/quadratic
# feat: File Browser Update (urls, examples, etc)

Resolves #323 

Functionality / things to test:

- Routing:
  - `/` loads the file menu _unless_ it's the first visit, then it loads the default file
  - `/?local={valid-file-id}` loads a file from memory
  - `/?local={invalid-text}` loads the file menu with an error
    - e.g. `/?local=foo`
  - `/?file={valid-url}` loads the `.grid` file over the network
    - Can test using example files from the `v0.3.0-beta` pulled from jsdeliver:
      - [/?file=https://cdn.jsdelivr.net/gh/quadratichq/quadratic@v0.3.0-beta/public/examples/python.grid](https://pr-308.de5w5iglj13on.amplifyapp.com/?file=https://cdn.jsdelivr.net/gh/quadratichq/quadratic@v0.3.0-beta/public/examples/python.grid)
      - [/?file=https://cdn.jsdelivr.net/gh/quadratichq/quadratic@v0.3.0-beta/public/examples/startup_portfolio.grid](https://pr-308.de5w5iglj13on.amplifyapp.com/?file=https://cdn.jsdelivr.net/gh/quadratichq/quadratic@v0.3.0-beta/public/examples/startup_portfolio.grid)
      - [/?file=https://cdn.jsdelivr.net/gh/quadratichq/quadratic@v0.3.0-beta/public/examples/expenses.grid](https://pr-308.de5w5iglj13on.amplifyapp.com/?file=https://cdn.jsdelivr.net/gh/quadratichq/quadratic@v0.3.0-beta/public/examples/expenses.grid)
  - `/?file={invalid-url}` loads the file menu with an error
    - e.g. `/?file=https://google.com`
- File menu:
  - Opens with: `CMD` + `O`, Toolbar "File -> Open...", or Command Palette "File: Open..."
  - Close with:  `ESC` or `X` in UI
  - Create a new file (names it "Untitled", goes to origin, zooms to 100%)
  - Open an example file (names it file name, goes to origin, zooms to 100%)
  - Open a file from local device (names it based on file on disk sans extension, goes to origin, zooms to 100%)
    - **Make sure to test this with an  old file** (something from the current version of production)
  - Open a file from a URL (names based on file path in, goes to origin, zooms to 100%)
    - Test with a valid `.grid` URL: https://cdn.jim-nielsen.com/files/sample-file.grid
    - Test with a valid `.grid` URL **from an old file**
    - Test with an invalid URL, e.g. https://google.com
  - Delete an in-memory file (deletes it from UI and indexDB)
  - Download a local copy of an in-memory file
- "File -> Save local copy" (from toolbar or command palette) downloads file with current name + `.grid` extension
- "File -> New" (from toolbar or command palette) starts a new file, zoomed to 100% at (0,0)
- In-app link to "Files" page in docs works
- Can rename a file from the top bar (empty filenames or filenames with only spaces `   ` are not allowed)
- "Open file" menu accessible from tablet/mobile
- Files you currently have in memory are migrated to our new storage mechanism
  - You can test this locally by running `main` on `localhost:3000` and create a couple different files in memory, then checkout this branch and load the app and the files you created should be accessible in the file menu

Todos:
- [x] Current files in memory should survive this new change when it merges
- [x] Write docs for how files/file formats work in Quadratic (code currently points at [docs.quadratichq.com/files](https://docs.quadratichq.com/files))
- [x] Fix contextual TODOs in code
- [x] Firing twice on each save (appears to be something with pixiapp, not new code)
- [x] Support old files
- [ ] Fix/update sample files
  - [ ] Update with version "1.1" meta info?
  - [x] Some aren't loading right now?
  - [x] Updates from `main`?
- [x] Add filename to document.title `{File Name} - Quadratic`
- [x] File issue for after merge: open separate PR to use context on `app` and `sheetConroller`

Nice to haves:
- [x] Example files that, when you clone them, appear with the same name as in the example file list
- [x] Weird flicker when sheet loads first, then open menu renders
  - Coming from how the app loads. `useLocalFiles` hook has an async action to load the file from memory. Would require re-architecting how this works...

Leaving out:

- Ability to deep link to specific tabs of the "Open file" menu, e.g. from the header click "File -> New from URL" which would open this new screen with the "Remote" URL tab selected
- Contexts for `sheetController` and `app` https://github.com/quadratichq/quadratic/issues/387
